File: Lista/Go/Q7.go
package main

import "fmt"

type Professor struct {
    Nome    string
    Escolas []*Escola
}

func (p Professor) ExibirEscolas() {
    fmt.Printf("Professor: %s\nLeciona nas seguintes escolas:\n", p.Nome)
    for _, escola := range p.Escolas {
        fmt.Printf("- %s\n", escola.Nome)
    }
}

func (p *Professor) AdicionarEscola(escola *Escola) {
    p.Escolas = append(p.Escolas, escola)
}

type Escola struct {
    Nome       string
    Professores []*Professor
}

func (e Escola) ExibirProfessores() {
    fmt.Printf("Escola: %s\nProfessores:\n", e.Nome)
    for _, professor := range e.Professores {
        fmt.Printf("- %s\n", professor.Nome)
    }
}

func (e *Escola) AdicionarProfessor(professor *Professor) {
    e.Professores = append(e.Professores, professor)
    professor.AdicionarEscola(e)
}

func main() {
    professor1 := &Professor{Nome: "Professor A"}
    professor2 := &Professor{Nome: "Professor B"}

    escola1 := &Escola{Nome: "Escola Alpha"}
    escola2 := &Escola{Nome: "Escola Beta"}

    escola1.AdicionarProfessor(professor1)
    escola1.AdicionarProfessor(professor2)

    escola2.AdicionarProfessor(professor1)

    escola1.ExibirProfessores()
    escola2.ExibirProfessores()

    professor1.ExibirEscolas()
    professor2.ExibirEscolas()
}

File: Lista/Python/Q7.py
class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []

    def adicionar_escola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)

    def exibir_escolas(self):
        escolas_nomes = [escola.nome for escola in self.escolas]
        return f'Professor {self.nome} leciona nas escolas: {", ".join(escolas_nomes)}'

class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []

    def adicionar_professor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)
            professor.adicionar_escola(self)

    def exibir_professores(self):
        professores_nomes = [professor.nome for professor in self.professores]
        return f'Escola {self.nome} tem os professores: {", ".join(professores_nomes)}'


escola1 = Escola('Escola A')
escola2 = Escola('Escola B')

professor1 = Professor('Carlos')
professor2 = Professor('Maria')

escola1.adicionar_professor(professor1)
escola1.adicionar_professor(professor2)

escola2.adicionar_professor(professor1)

print(escola1.exibir_professores())
print(escola2.exibir_professores())

print(professor1.exibir_escolas())
print(professor2.exibir_escolas())